src/darwin/CHIPTool/CHIPTool/View Controllers/Echo client/EchoViewController.m

```diff
@@ -0,0 +1,225 @@
+    [sendMessageView.trailingAnchor constraintEqualToAnchor:_stackView.trailingAnchor].active = YES;
+    
+    // Result message
+    _resultLabel = [UILabel new];
+    _resultLabel.hidden = YES;
+    _resultLabel.font = [UIFont systemFontOfSize:17];
+    _resultLabel.textColor = UIColor.systemBlueColor;
+    _resultLabel.lineBreakMode = NSLineBreakByWordWrapping;
+    _resultLabel.numberOfLines = 0;
+    [_stackView addArrangedSubview:_resultLabel];
+    
+    _resultLabel.translatesAutoresizingMaskIntoConstraints = false;
+    [_resultLabel.trailingAnchor constraintEqualToAnchor:_stackView.trailingAnchor].active = YES;
+    _resultLabel.adjustsFontSizeToFitWidth = YES;
+}
+
+- (void)updateResult:(NSString *)result
+{
+    _resultLabel.hidden = NO;
+    _resultLabel.text = result;
+}
+
+// MARK: UIButton actions
+
+- (IBAction)saveServerID:(id)sender
+{
+    NSString * serverIP = [self.serverIPTextField text];
+    if ([serverIP length] > 0) {
+        CHIPSetDomainValueForKey(kCHIPToolDefaultsDomain, kIPKey, serverIP);
+    }
+}
+
+- (IBAction)sendMessage:(id)sender
+{
+    NSString * msg = [self.messageTextField text];
+    if (msg.length == 0) {
+        msg = [self.messageTextField placeholder];
+    }
+
+    [self.connectivityManager reconnectIfNeeded];
+
+    // send message
+    if ([self.chipController isConnected]) {
+        NSError * error;
+        BOOL didSend = [self.chipController sendMessage:[msg dataUsingEncoding:NSUTF8StringEncoding] error:&error];
+        if (!didSend) {
+            NSString *errorString = [@"Error: " stringByAppendingString:error.localizedDescription];
+            [self updateResult:errorString];
+        } else {
+            [self updateResult:@"Message Sent"];
+        }
+    } else {
+        [self updateResult:@"Controller not connected"];
+    }
+}
+
+// MARK: CHIPDeviceControllerDelegate
+- (void)deviceControllerOnConnected {
+    NSLog(@"Status: Device connected");
+}
+
+- (void)deviceControllerOnError:(nonnull NSError *)error {
+    NSLog(@"Status: Device Controller error %@", [error description]);
+    if (error) {
+       NSString *stringError = [@"Error: " stringByAppendingString:error.description];
+       dispatch_after(dispatch_time(DISPATCH_TIME_NOW, 5.0), dispatch_get_main_queue(), ^{
+           [self updateResult:stringError];
+       });
+    }
+    
+}
+
+- (void)deviceControllerOnMessage:(nonnull NSData *)message {
+    NSString *stringMessage;
+    if ([CHIPDeviceController isDataModelCommand:message] == YES) {
+        stringMessage = [CHIPDeviceController commandToString:message];
+    } else {
+        stringMessage = [[NSString alloc] initWithData:message encoding:NSUTF8StringEncoding];
+    }
+    NSString *resultMessage = [@"Echo Response: " stringByAppendingFormat:@"%@", stringMessage];
+    dispatch_after(dispatch_time(DISPATCH_TIME_NOW, 5.0), dispatch_get_main_queue(), ^{
+        [self updateResult:resultMessage];
+    });
+}
+
+// MARK: CHIPConnectivityManagerDelegate
```

Comment:
did you mean `#pragma mark` or does this end up doing the same thing?

Reply:
// MARK: is swift and objc compatible vs pragma mark is for objective c 

I always advised to use MARK on my old team to keep consistent between the objc/swift files so I keep doing that

Reply:
TIL, thanks!